projects/RabbitMQ.Client/client/impl/RabbitMQActivitySource.cs

```diff
@@ -0,0 +1,166 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Text;
+using System.Threading;
+using RabbitMQ.Client.Events;
+using RabbitMQ.Client.Framing.Impl;
+
+namespace RabbitMQ.Client
+{
+    public static class RabbitMQActivitySource
+    {
+        internal const string MessageId = "messaging.message.id";
+        internal const string MessageConversationId = "messaging.message.conversation_id";
+        internal const string MessagingOperation = "messaging.operation";
+        internal const string MessagingSystem = "messaging.system";
+        internal const string MessagingDestination = "messaging.destination.name";
+        internal const string MessagingDestinationRoutingKey = "messaging.rabbitmq.destination.routing_key";
+        internal const string MessagingBodySize = "messaging.message.body.size";
+        internal const string MessagingEnvelopeSize = "messaging.message.envelope.size";
+        internal const string ProtocolName = "network.protocol.name";
+        internal const string ProtocolVersion = "network.protocol.version";
+        internal const string RabbitMQDeliveryTag = "messaging.rabbitmq.delivery.tag";
+
+        internal static readonly ActivitySource s_source = new ActivitySource(Name, typeof(RabbitMQActivitySource).Assembly.GetCustomAttribute<AssemblyInformationalVersionAttribute>().InformationalVersion);
+        public const string Name = "RabbitMQ.Client";
+        public static bool UseRoutingKeyAsOperationName { get; set; } = true;
+
+        internal static Activity Send(string routingKey, string exchange, int bodySize)
+        {
+            if (s_source.HasListeners())
+            {
+                Activity activity = StartRabbitMQActivity(UseRoutingKeyAsOperationName ? $"{routingKey} publish" : "publish", ActivityKind.Producer);
+                if (activity?.IsAllDataRequested == true)
+                {
+                    PopulateMessagingTags("publish", routingKey, exchange, 0, null, bodySize, activity);
+                }
+
+                return activity;
+            }
+
+            return null;
+        }
+
+        internal static Activity Receive(string routingKey, string exchange, ulong deliveryTag, IReadOnlyBasicProperties readOnlyBasicProperties, int bodySize)
```

Comment:
The IReadOnlyBasicProperties boxes any struct property type. :(

Reply:
I'll see if I can replace/overload it with more concrete types

Reply:
Changed the code a bit to just send the .Headers dictionary instead of having to copy the struct. Also made the IReadOnlyBasicProperties parameter an `in` parameter in the one place I have it since it's not an async method :)